File: sgu-kdzr-server/src/main/java/utez/edu/mx/sgukdzrserver/modules/person/PersonController.java
package utez.edu.mx.sgukdzrserver.modules.person;

import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;
import utez.edu.mx.sgukdzrserver.modules.person.dto.PersonRequest;
import utez.edu.mx.sgukdzrserver.modules.person.dto.PersonResponse;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/persons")
@CrossOrigin(origins = "*")
public class PersonController {

    private final PersonService personService;

    @Autowired
    public PersonController(PersonService personService) {
        this.personService = personService;
    }

    /**
     * Obtener todas las personas
     * GET /api/persons
     */
    @GetMapping
    public ResponseEntity<Map<String, Object>> getAllPersons() {
        try {
            List<PersonResponse> persons = personService.getAllPersons();
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Personas obtenidas exitosamente");
            response.put("data", persons);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Error al obtener las personas: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }

    /**
     * Obtener una persona por ID
     * GET /api/persons/{id}
     */
    @GetMapping("/{id}")
    public ResponseEntity<Map<String, Object>> getPersonById(@PathVariable Long id) {
        try {
            PersonResponse person = personService.getPersonById(id);
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Persona encontrada");
            response.put("data", person);
            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Error al obtener la persona: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }

    /**
     * Crear una nueva persona
     * POST /api/persons
     */
    @PostMapping
    public ResponseEntity<Map<String, Object>> createPerson(
            @Valid @RequestBody PersonRequest request,
            BindingResult bindingResult) {

        if (bindingResult.hasErrors()) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Errores de validación");
            response.put("errors", bindingResult.getFieldErrors()
                    .stream()
                    .map(error -> Map.of(
                            "field", error.getField(),
                            "message", error.getDefaultMessage()
                    ))
                    .collect(Collectors.toList()));
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        try {
            PersonResponse person = personService.createPerson(request);
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Persona creada exitosamente");
            response.put("data", person);
            return ResponseEntity.status(HttpStatus.CREATED).body(response);
        } catch (RuntimeException e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", e.getMessage());
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Error al crear la persona: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }

    /**
     * Actualizar una persona existente
     * PUT /api/persons/{id}
     */
    @PutMapping("/{id}")
    public ResponseEntity<Map<String, Object>> updatePerson(
            @PathVariable Long id,
            @Valid @RequestBody PersonRequest request,
            BindingResult bindingResult) {

        if (bindingResult.hasErrors()) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Errores de validación");
            response.put("errors", bindingResult.getFieldErrors()
                    .stream()
                    .map(error -> Map.of(
                            "field", error.getField(),
                            "message", error.getDefaultMessage()
                    ))
                    .collect(Collectors.toList()));
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        try {
            PersonResponse person = personService.updatePerson(id, request);
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Persona actualizada exitosamente");
            response.put("data", person);
            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", e.getMessage());
            HttpStatus status = e.getMessage().contains("no encontrada")
                    ? HttpStatus.NOT_FOUND
                    : HttpStatus.BAD_REQUEST;
            return ResponseEntity.status(status).body(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Error al actualizar la persona: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }

    /**
     * Eliminar una persona
     * DELETE /api/persons/{id}
     */
    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, Object>> deletePerson(@PathVariable Long id) {
        try {
            personService.deletePerson(id);
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Persona eliminada exitosamente");
            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Error al eliminar la persona: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }

    /**
     * Buscar persona por email
     * GET /api/persons/email/{email}
     */
    @GetMapping("/email/{email}")
    public ResponseEntity<Map<String, Object>> getPersonByEmail(@PathVariable String email) {
        try {
            PersonResponse person = personService.getPersonByEmail(email);
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Persona encontrada");
            response.put("data", person);
            return ResponseEntity.ok(response);
        } catch (RuntimeException e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
        } catch (Exception e) {
            Map<String, Object> response = new HashMap<>();
            response.put("success", false);
            response.put("message", "Error al obtener la persona: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }
}

File: sgu-kdrz-client/src/modules/person/person.controller.js
const PersonController = {};
const ENV = import.meta.env;

const API_URL = `http://${ENV.VITE_API_HOST}:${ENV.VITE_API_PORT}${ENV.VITE_API_BASE}`;

/**
 * Obtener todas las personas
 */
PersonController.getAll = async () => {
    try {
        const response = await fetch(`${API_URL}/persons`, {
            method: 'GET',
            headers: {
                'Accept': 'application/json'
            }
        });
        const result = await response.json();
        return result;
    } catch (error) {
        console.error('Error al obtener personas:', error);
        throw error;
    }
}

/**
 * Obtener una persona por ID
 */
PersonController.getById = async (id) => {
    try {
        const response = await fetch(`${API_URL}/persons/${id}`, {
            method: 'GET',
            headers: {
                'Accept': 'application/json'
            }
        });
        const result = await response.json();
        return result;
    } catch (error) {
        console.error('Error al obtener persona:', error);
        throw error;
    }
}

/**
 * Crear una nueva persona
 */
PersonController.create = async (personData) => {
    try {
        const response = await fetch(`${API_URL}/persons`, {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(personData)
        });
        const result = await response.json();
        return result;
    } catch (error) {
        console.error('Error al crear persona:', error);
        throw error;
    }
}

/**
 * Actualizar una persona existente
 */
PersonController.update = async (id, personData) => {
    try {
        const response = await fetch(`${API_URL}/persons/${id}`, {
            method: 'PUT',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(personData)
        });
        const result = await response.json();
        return result;
    } catch (error) {
        console.error('Error al actualizar persona:', error);
        throw error;
    }
}

/**
 * Eliminar una persona
 */
PersonController.delete = async (id) => {
    try {
        const response = await fetch(`${API_URL}/persons/${id}`, {
            method: 'DELETE',
            headers: {
                'Accept': 'application/json'
            }
        });
        const result = await response.json();
        return result;
    } catch (error) {
        console.error('Error al eliminar persona:', error);
        throw error;
    }
}

export default PersonController;
